src/server/ua_services_attribute.c

```diff
@@ -1809,6 +1913,91 @@ copyAttributeIntoNode(UA_Server *server, UA_Session *session,
     return UA_STATUSCODE_GOOD;
 }
 
+#if UA_MULTITHREADING >= 100
+static void
+Operation_WriteAsync(UA_Server *server, UA_Session *session, UA_UInt32 requestId,
+                    UA_UInt32 requestHandle, size_t opIndex,
+                    UA_WriteRequest *opRequest, UA_StatusCode *opResult,
+                    UA_AsyncResponse **ar) {
+    UA_assert(session != NULL);
+
+    /* Get the node (with only the selected attribute if the NodeStore supports that) */
+    const UA_Node *node =
+        UA_NODESTORE_GET_SELECTIVE(server, &opRequest->nodesToWrite[opIndex].nodeId,
+                                   attributeId2AttributeMask((UA_AttributeId) opRequest->nodesToWrite[opIndex].attributeId),
+                                   UA_REFERENCETYPESET_NONE,
+                                   UA_BROWSEDIRECTION_INVALID);
+    if(!node){
+        *opResult = UA_STATUSCODE_BADNODEIDUNKNOWN;
+        return;
+    }
+
+    //Synchronous execution
+    if(!node->variableNode.async) {
+        *opResult = UA_Server_editNode(server, session, &opRequest->nodesToWrite[opIndex].nodeId,
+                                     (UA_EditNodeCallback)copyAttributeIntoNode,
+                                     (void*)(uintptr_t) &opRequest->nodesToWrite[opIndex]);
+    }
+
+     /* Async Write No AsyncResponse allocated so far */
+    if(!*ar) {
+        *opResult =
+            UA_AsyncManager_createAsyncResponse(&server->asyncManager, server,
+                                                &session->sessionId, requestId, requestHandle,
+                                                UA_ASYNCOPERATIONTYPE_WRITE, ar);
+        if(*opResult != UA_STATUSCODE_GOOD)
+        goto cleanup;
+    }
+
+    //Create the Async Request to be taken by workers
+    //TODO make "UA_AsyncManager_createAsyncOp" generic for read/write
```

Comment:
I wonder if the todo is still valid or is in the wrong place?
If UA_AsyncManager_createAsyncOp should somehow become more generic, then I would expect the todo comment to be in the function.

Reply:
Right, old artefact. Removed.